packages/composites/token-allocation/src/endpoint/index.ts

```diff
@@ -0,0 +1,181 @@
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      return acc.add(coins.mul(val!))
+    }, new Decimal(0))
+    .toNumber()
+}
+
+export const marketCapTotalValue = (
+  source: string,
+  allocations: TokenAllocations,
+  quote: string,
+  data: ResponsePayload,
+): number => {
+  return allocations
+    .reduce((acc, t) => {
+      const val = data[t.symbol].quote[quote].marketCap
+      if (!val)
+        throw new Error(
+          `ERROR: No marketcap value found for ${t.symbol}/${quote} from the ${source} adapter.`,
+        )
+      const coins = new Decimal(t.balance.toString(10)).div(10 ** t.decimals)
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      return acc.add(coins.mul(val!))
+    }, new Decimal(0))
+    .toNumber()
+}
+
+const toValidAllocations = (allocations: any[]): TokenAllocations => {
+  const _toValidSymbol = (symbol: string) => {
+    if (!symbol)
+      throw new AdapterError({ message: `Symbol not available for all tokens.`, statusCode: 400 })
+    return symbol.toUpperCase()
+  }
+  const _toValidDecimals = (decimals: number | undefined) => {
+    if (decimals === undefined) return DEFAULT_TOKEN_DECIMALS
+    return Number.isInteger(decimals) && decimals >= 0 ? decimals : DEFAULT_TOKEN_DECIMALS
+  }
+  const _toValidBalance = (balance: number | string | undefined, decimals: number) => {
+    if (!balance) return DEFAULT_TOKEN_BALANCE * 10 ** decimals
+    let BNbalance
+    try {
+      BNbalance = BigNumber.from(balance.toString())
+    } catch (e) {
+      throw new AdapterError({ message: `Invalid balance: ${e.message}`, statusCode: 400 })
+    }
+    if (BNbalance.isNegative())
+      throw new AdapterError({ message: `Balance cannot be negative`, statusCode: 400 })
+    return balance
+  }
+  return allocations.map((t: any) => {
+    const decimals = _toValidDecimals(t.decimals)
+    return {
+      symbol: _toValidSymbol(t.symbol),
+      decimals,
+      balance: _toValidBalance(t.balance, decimals),
+    }
+  })
+}
+
+const computePrice = async (
+  source: string,
+  getPrices: GetPrices,
+  allocations: TokenAllocations,
+  quote: string,
+  additionalInput: Record<string, unknown>,
+) => {
+  const symbols = (allocations as TokenAllocations).map((t) => t.symbol)
+  const payload = await getPrices(symbols, quote, additionalInput, false)
+  const result = priceTotalValue(source, allocations, quote, payload)
+  return { payload, result }
+}
+
+const computeMarketCap = async (
+  source: string,
+  getPrices: GetPrices,
+  allocations: TokenAllocations,
+  quote: string,
+  additionalInput: Record<string, unknown>,
+) => {
+  const symbols = (allocations as TokenAllocations).map((t) => t.symbol)
+  const payload = await getPrices(symbols, quote, additionalInput, true)
+
+  const result = marketCapTotalValue(source, allocations, quote, payload)
+  return { payload, result }
+}
```

Comment:
Since these methods are almost identical, can you merge them? You would need to pass in a string for the field to access, like `valueField: 'price' | 'marketcap'`, but IMO it would look better condensed.

Reply:
This was just copy pasted moving it to its own file to clean up having `adapter.ts`, but yeah I can change that when I get time to re-do the versions on this PR.

Reply:
Didn't clean this up to maintain previous functionality.